src/main/java/io/cdap/plugin/batch/source/BSONObjectToRecordTransformer.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.plugin.batch.source;
+
+import com.google.common.collect.Lists;
+import io.cdap.cdap.api.data.format.StructuredRecord;
+import io.cdap.cdap.api.data.format.UnexpectedFormatException;
+import io.cdap.cdap.api.data.schema.Schema;
+import org.bson.BSONObject;
+import org.bson.BsonUndefined;
+import org.bson.types.BasicBSONList;
+import org.bson.types.Decimal128;
+import org.bson.types.ObjectId;
+
+import java.math.BigDecimal;
+import java.nio.ByteBuffer;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Transforms {@link BSONObject} to {@link StructuredRecord}.
+ */
+public class BSONObjectToRecordTransformer {
+
+  private final Schema schema;
+
+  public BSONObjectToRecordTransformer(Schema schema) {
+    this.schema = schema;
+  }
+
+  /**
+   * Transforms given {@link BSONObject} to {@link StructuredRecord}.
+   *
+   * @param bsonObject BSON object to be transformed.
+   * @return {@link StructuredRecord} that corresponds to the given {@link BSONObject}.
+   */
+  public StructuredRecord transform(BSONObject bsonObject) {
+    return convertRecord(bsonObject, schema);
+  }
+
+  private StructuredRecord convertRecord(Object object, Schema schema) {
+    BSONObject bsonObject = (BSONObject) object;
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    List<Schema.Field> fields = Objects.requireNonNull(schema.getFields(), "Schema fields cannot be empty");
+    for (Schema.Field field : fields) {
+      Object value = extractValue(bsonObject.get(field.getName()), field.getSchema());
+      Schema.LogicalType fieldLogicalType = field.getSchema().isNullable()
+        ? field.getSchema().getNonNullable().getLogicalType()
+        : field.getSchema().getLogicalType();
+
+      if (fieldLogicalType == null) {
+        // Set values of simple types
+        builder.set(field.getName(), value);
+        continue;
+      }
+
+      // Set values of logical types properly
+      switch (fieldLogicalType) {
```

Comment:
instead of extracting a value then checking the schema here and casting, it would be better to make 'extractValue' 'setValue' and pass in the builder.

Reply:
The problem is that `extractValue` can be used to recursively convert values of Arrays and Maps:
```
  private Object convertMap(String fieldName, BSONObject bsonObject, Schema schema) {
    Schema valueSchema = schema.getMapSchema().getValue().isNullable()
      ? schema.getMapSchema().getValue().getNonNullable() : schema.getMapSchema().getValue();
    Map<String, Object> extracted = new HashMap<>();
    for (String key : bsonObject.keySet()) {
      Object value = bsonObject.get(key);
      extracted.put(key, extractValue(fieldName, value, valueSchema));
    }
    return extracted;
  }
	
  private Object convertArray(String fieldName, BasicBSONList bsonList, Schema schema) {
    List<Object> values = Lists.newArrayListWithCapacity(bsonList.size());
    Schema componentSchema = schema.getComponentSchema().isNullable() ? schema.getComponentSchema().getNonNullable()
      : schema.getComponentSchema();
    for (Object obj : bsonList) {
      values.add(extractValue(fieldName, obj, componentSchema));
    }
    return values;
  }
```

whereas the builder can be used only when converting a `record`. 

Thus it seems that we have to check field logical types in two places: 
1. When converting the BSON value, which not necessarily will be set on a structured record
2. When setting the value on a structured record

> `convertMap` method was missing. I created it along with the corresponding [testTransformComplexMap](testTransformComplexMap) test

Reply:
hm I see. Though one problem with this is that logical types like timestamp and decimal are expected to be stored as their physical types in StructuredRecord. So it's actually incorrect for extractValue() to return ZonedDateTime and BigDecimal, since we need the collection to contain the physical representation.

What we need to do is change extractValue to always return the physical type, which also removes the need to check the type in this method.

This is all more complicated than it really should be, I've opened https://issues.cask.co/browse/CDAP-15727 to see if we can make this easier in the future.

Reply:
Done.